st2common/tests/unit/test_operators.py

```diff
@@ -19,16 +19,540 @@
 from st2common.util import date as date_utils
 
 
+def dict_strict_equal(d1, d2):
+    """
+    Compare two dictionaries, ensuring that all keys exist in both and that the
+    values for each key are the same.
+    """
+    if len(d1.keys()) != len(d2.keys()):
+        return False
+    for k1, v1 in d1.items():
+        if k1 not in d2 or d2[k1] != v1:
+            return False
+    return True
+
+
+def list_of_dicts_strict_equal(lofd1, lofd2):
+    """
+    Ensure that two unordered lists contain the same dicts
+    """
+    t2 = list(lofd2)  # mutable copy
+    if len(lofd1) != len(lofd2):
+        return False
+    for i1, el1 in enumerate(lofd1):
+        for i2, el2 in enumerate(t2):
+            if dict_strict_equal(el1, el2):
```

Comment:
I'm fine with this approach, but usually you need to be careful when dealing with mutable copies and iterating over them at the same time because the result might not always be what you expect.

So not a blocker, but I would prefer an approach where the list doesn't need to be mutated in place during iteration.

Reply:
I just noticed those functions are only used in the tests so it's even less big of a deal.

Reply:
Yep, I don't like it either, so I made sure the semantics all worked out. I even wrote test cases for the comparison functions I use in the tests just to make sure they worked the way I expected them to work. 😆